### PR TITLE
added gmatmul

### PR DESCRIPTION
**Pull request by SteveDiamond**
I added the gmatmul atom discussed in #999 . I need some help though, because when I make the docs no API entry is created for gmatmul.

### CODE REVIEW

**Comment by SteveDiamond** [comment]
Also @akshayka do I need to add some DPP logic?

File: cvxpy/atoms/gmatmul.py [hunk 1]
```diff
@@ -0,0 +1,120 @@
+"""
+Copyright 2013 Steven Diamond
```
Comment (akshayka) [inline cvxpy/atoms/gmatmul.py:2]: Should this be "Copyright, the CVXPY authors", as in https://github.com/cvxgrp/cvxpy/blob/master/copyright.txt?

File: doc/source/tutorial/dgp/index.rst [hunk 2]
```diff
@@ -591,6 +592,14 @@ and returns a vector or matrix.
      - |convex| log-log convex
      - |incr| incr.
 
+   * - :ref:`gmatmul(A, x) <gmatmul>`
+
+       :math:`A \in \mathbf{R}^{m \times n}`
+     - :math:`\left[\begin{matrix}\prod_{j=1}^n x_j^{A_{1j}} \\\vdots \\\prod_{j=1}^n x_j^{A_{mj}}\end{matrix}\right]`
+     - :math:`x \in \mathbf{R}^n_{++}`
+     - |affine| log-log affine
+     - None
```
Comment (akshayka) [inline doc/source/tutorial/dgp/index.rst:601]: This atom is increasing in x, right? (Since its domain is x positive?)

**Comment by akshayka** [comment]
> Also @akshayka do I need to add some DPP logic?

Yes, some DPP logic will be necessary.

I haven't merged the DPP for DGP branch (dgp-dpp) into master yet. I think the branch is ready for review (or will very soon be ready for review), and all the tests are passing.

If this change gets merged before dgp-dpp is merged, I can rebase and add the DPP logic.

**Comment by akshayka** [comment]
> I need some help though, because when I make the docs no API entry is created for gmatmul.

You'll need to modify the appropriate file in

https://github.com/cvxgrp/cvxpy/tree/master/doc/source/api_reference

to tell Sphinx to generate an API entry in the website docs.

File: doc/source/tutorial/dgp/index.rst (hunk 2 above)
Comment (SteveDiamond) [reply doc/source/tutorial/dgp/index.rst:601]: No, because you could for example have $A = -1$.

Comment (SteveDiamond) [reply doc/source/tutorial/dgp/index.rst:601]: The actual logic of the atom is increasing if A is all non-negative, decreasing if A is all non-positive. But that seemed to involved for the tutorial description.

**Comment by SteveDiamond** [comment]
> > I need some help though, because when I make the docs no API entry is created for gmatmul.
> 
> You'll need to modify the appropriate file in
> 
> https://github.com/cvxgrp/cvxpy/tree/master/doc/source/api_reference
> 
> to tell Sphinx to generate an API entry in the website docs.

That worked. How do I make the DGP tutorial entry link to it though?

File: cvxpy/atoms/gmatmul.py (hunk 1 above)
Comment (SteveDiamond) [reply cvxpy/atoms/gmatmul.py]: Thanks! This is fixed.

**Comment by SteveDiamond** [comment]
I got the linking from the DGP atoms page to work (just needed to delete the old build files before building html again). Anyway, from my point of view the PR is done.

File: doc/source/tutorial/dgp/index.rst (hunk 2 above)
Comment (akshayka) [reply doc/source/tutorial/dgp/index.rst:601]: Oh, right. I think we could include that fact in the documentation. For example, the documentation on power() for DGP says incr when p > 0, decr when p < 0. If there's a discrepancy between what's documented and what's implemented, that might lead to confusion.

Comment (SteveDiamond) [reply doc/source/tutorial/dgp/index.rst]: Ok I added this logic to the docs.

**Comment by SteveDiamond** [comment]
Thanks!

**Pull request merged**